src/Common/MoistThermodynamics/profiles.jl

```diff
@@ -0,0 +1,284 @@
+    "temperature (K)"
+    T::FT
+end
+
+function (profile::IsothermalProfile)(
+    param_set,
+    e_pot::FT,
+) where {FT<:AbstractFloat}
+    _R_d::FT = R_d(param_set)
+    _MSLP::FT = MSLP(param_set)
+
+    # Pressure
+    p = _MSLP * exp(-e_pot / (_R_d * profile.T))
+
+    # Density
+    ρ = p / (_R_d * T)
+    return profile.T, p, ρ
+end
+
+"""
+    DryAdiabaticProfile{F} <: AbstractReferenceState
+
+
+A temperature profile that has uniform potential temperature θ until it
+reaches the minimum specified temperature `T_min`
+
+# Fields
+
+$(DocStringExtensions.FIELDS)
+"""
+struct DryAdiabaticProfile{FT} <: AbstractReferenceState{FT}
+    "minimum temperature (K)"
+    T_min::FT
+    "potential temperature (K)"
+    θ::FT
+end
+
+function (profile::DryAdiabaticProfile)(
+    param_set,
+    z::FT,
+) where {FT<:AbstractFloat}
+    _cp_d::FT = cp_d(param_set)
+    _grav::FT = grav(param_set)
+    temperature_profile = LinearTemperatureProfile(
+        profile.T_min,
+        profile.θ,
+        FT(_grav / _cp_d))
+
+    return temperature_profile(
+        param_set,
+        z,
+    )
+end
+
+"""
+    LinearTemperatureProfile{F} <: AbstractReferenceState
+
+A temperature profile which decays linearly with height `z`, until it reaches a minimum specified temperature.
+
+```math
+T(z) = \\max(T_{\\text{surface}} − Γ z, T_{\\text{min}})
+```
+
+# Fields
+
+$(DocStringExtensions.FIELDS)
+"""
+struct LinearTemperatureProfile{FT} <: AbstractReferenceState{FT}
+    "minimum temperature (K)"
+    T_min::FT
+    "surface temperature (K)"
+    T_surface::FT
+    "lapse rate (K/m)"
+    Γ::FT
+end
+
+function (profile::LinearTemperatureProfile)(
+    param_set::AbstractParameterSet,
+    z::FT,
+) where {FT<:AbstractFloat}
+
+    _R_d::FT = R_d(param_set)
+    _grav::FT = grav(param_set)
+    _MSLP::FT = MSLP(param_set)
+
+    # Temperature
+    T = max(profile.T_surface - profile.Γ * z, profile.T_min)
+
+    # Pressure
+    p = _MSLP * (T / profile.T_surface)^(_grav / (_R_d * profile.Γ))
+    if T == profile.T_min
+        z_top = (profile.T_surface - profile.T_min) / profile.Γ
+        H_min = _R_d * profile.T_min / _grav
+        p *= exp(-(z - z_top) / H_min)
+    end
+
+    # Density
+    ρ = p / (_R_d * T)
+    return (T, p, ρ)
+end
+
+"""
+    tested_profiles(param_set, n::Int, ::Type{FT})
+
+A range of input arguments to thermodynamic state constructors
+ - `param_set` an `AbstractParameterSet`, see the [`MoistThermodynamics`](@ref) for more details
+ - `e_int` internal energy
+ - `ρ` (moist-)air density
+ - `q_tot` total specific humidity
+ - `q_pt` phase partition
+ - `T` air temperature
+ - `θ_liq_ice` liquid-ice potential temperature
+that are tested for convergence in saturation adjustment.
+
+Note that the output vectors are of size ``n*n_RH``, and they
+should span the input arguments to all of the constructors.
+"""
+function tested_profiles(param_set::APS, n::Int, test_profile::AbstractReferenceState{FT}) where {FT}
```

Comment:
@tapios can you review this function? I took RH to be in the range [0, 1]. I tried increasing it beyond 1 but saturation adjustment started failing. Maybe we can do an interactive debug session soon. Also, I don't know that this function is valid for any `test_profile`. At the very least, this is different than before because now total specific humidity is computed with `q_tot = RH .* q_vap_saturation.(Ref(param_set), T, ρ)`

Reply:
Testing saturation adjustment for RH <= 1 is kind of pointless because it will not need to iterate then. Probably first thing to do here is to update the convergence criteria in saturation adjustment to make them residual based.